### PR TITLE
Button cleanup

### PR DESCRIPTION
**Pull request by SharkyKZ**
Pull Request for Issue # .
### Summary of Changes
This cleans up mainly modal-related buttons. Most notable change is restoration of `<button>` element which was removed in PRs #11787, #14483, #16156. The original issue #11765 was caused by missing `type` attribute in user field layout. Although switching to `<a>` tag fixed the issue, it wasn't the right way to do it.

For future reference:
Use `<button>` element in favor of `<a>` for button functionality whenever possible (https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_button_role).
Always add `type` attribute (most browsers default to `type="submit"` without it).
`button` is not a valid value for `type` attribute on `<a>` (https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-type).
Don't use `aria-hidden="true"` on buttons containing readable text. Use it on non-readable elements like icons.

### Testing Instructions
1. Code review.
2. Check that buttons work correctly.
3. Confirm that you cannot reproduce issue #11765.
### Documentation Changes Required
No.

### CODE REVIEW

File: administrator/components/com_categories/models/fields/modal/category.php [hunk 1]
```diff
@@ -129,11 +129,11 @@ function jSelectCategory_" . $this->id . "(id, title, object) {
 		if ($allowSelect)
 		{
 			$html .= '<a'
```
Comment (Quy) [inline administrator/components/com_categories/models/fields/modal/category.php]: Change.

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on a2aa52711dfec6da401c91d452273166823b5ef8<br /><br />tested all the modals I could think of
Tested to make sure the original bug had not come back<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20682">issues.joomla.org/tracker/joomla-cms/20682</a>.</sub>

**Comment by brianteeman** [comment]
Searching the entire codebase for 'a class="btn' reveals a lot more that might need to be checked<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20682">issues.joomla.org/tracker/joomla-cms/20682</a>.</sub>

File: layouts/joomla/modal/header.php [hunk 2]
```diff
@@ -34,7 +34,14 @@
...
 <div class="modal-header">
 	<?php if (!isset($params['closeButton']) || $params['closeButton']) : ?>
-		<button type="button" class="close novalidate" data-dismiss="modal">×</button>
+		<button
+			type="button"
+			class="close novalidate"
+			data-dismiss="modal"
+			aria-label="<?php echo JText::_('JLIB_HTML_BEHAVIOR_CLOSE'); ?>"
+			>
+			<span aria-hidden="true">×</span>
```
Comment (Quy) [inline layouts/joomla/modal/header.php]: Change `x` to `&times;`

File: administrator/templates/isis/html/layouts/joomla/form/field/user.php [hunk 3]
```diff
@@ -109,14 +109,21 @@
 	<div class="input-append">
 		<input <?php echo ArrayHelper::toString($inputAttributes); ?> readonly />
 		<?php if (!$readonly) : ?>
-			<a class="btn btn-primary button-select" title="<?php echo JText::_('JLIB_FORM_CHANGE_USER') ?>"><span class="icon-user"></span></a>
+			<button
+				type="button"
+				class="btn btn-primary button-select"
+				title="<?php echo JText::_('JLIB_FORM_CHANGE_USER') ?>"
```
Comment (Quy) [inline administrator/templates/isis/html/layouts/joomla/form/field/user.php]: Add `;` and below.

File: administrator/templates/hathor/html/com_menus/menus/default.php [hunk 4]
```diff
@@ -161,9 +161,9 @@
 										'title'  => JText::_('COM_MENUS_EDIT_MODULE_SETTINGS'),
 										'height' => '300px',
 										'width'  => '800px',
-										'footer' => '<a class="btn" type="button" data-dismiss="modal" aria-hidden="true">'
-											. JText::_('JLIB_HTML_BEHAVIOR_CLOSE') . '</a>'
-											. '<button class="btn btn-success" data-dismiss="modal" aria-hidden="true" onclick="jQuery(\'#module'
+										'footer' => '<button type="button" class="btn" data-dismiss="modal">'
+											. JText::_('JLIB_HTML_BEHAVIOR_CLOSE') . '</button>'
+											. '<button class="btn btn-success" data-dismiss="modal" onclick="jQuery(\'#module'
```
Comment (Quy) [inline administrator/templates/hathor/html/com_menus/menus/default.php]: Add `type`

File: administrator/components/com_contact/views/contacts/tmpl/default_batch.php [hunk 5]
```diff
@@ -13,7 +13,9 @@
 ?>
 <div class="modal hide fade" id="collapseModal">
 	<div class="modal-header">
-		<button type="button" class="close" data-dismiss="modal">&#215;</button>
+		<button type="button" class="close" data-dismiss="modal" aria-label="<?php echo JText::_('JLIB_HTML_BEHAVIOR_CLOSE'); ?>">
+			<span aria-hidden="true">&#215;</span>
```
Comment (Quy) [inline administrator/components/com_contact/views/contacts/tmpl/default_batch.php]: Change to `&times;`

File: administrator/components/com_contact/models/fields/modal/contact.php [hunk 6]
```diff
@@ -204,18 +204,18 @@ function jSelectContact_" . $this->id . "(id, title, object) {
 					'width'       => '800px',
 					'bodyHeight'  => '70',
 					'modalWidth'  => '80',
-					'footer'      => '<a role="button" class="btn" aria-hidden="true"'
+					'footer'      => '<button type="button" class="btn"'
 							. ' onclick="window.processModalEdit(this, \''
 							. $this->id . '\', \'add\', \'contact\', \'cancel\', \'contact-form\', \'jform_id\', \'jform_name\'); return false;">'
 							. JText::_('JLIB_HTML_BEHAVIOR_CLOSE') . '</a>'
```
Comment (Quy) [inline administrator/components/com_contact/models/fields/modal/contact.php]: Change to `</button>`

File: administrator/components/com_contact/models/fields/modal/contact.php [hunk 7]
```diff
@@ -204,18 +204,18 @@ function jSelectContact_" . $this->id . "(id, title, object) {
...
-							. '<a role="button" class="btn btn-primary" aria-hidden="true"'
+							. '<button type="button" class="btn btn-primary"'
 							. ' onclick="window.processModalEdit(this, \''
 							. $this->id . '\', \'add\', \'contact\', \'save\', \'contact-form\', \'jform_id\', \'jform_name\'); return false;">'
 							. JText::_('JSAVE') . '</a>'
```
Comment (Quy) [inline administrator/components/com_contact/models/fields/modal/contact.php]: Change to `</button>`

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 2dc43631d50510aaa52dd9f72847ae8eaaea671d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20682">issues.joomla.org/tracker/joomla-cms/20682</a>.</sub>

**Comment by Quy** [comment]
@brianteeman This PR cleans up mainly modal-related buttons.

**Comment by ghost** [comment]
@brianteeman can you please retest?

File: administrator/templates/hathor/html/com_templates/template/default.php [hunk 8]
```diff
@@ -274,7 +278,7 @@ function clearCoords()
 					</div>
 				</div>
 				<div class="modal-footer">
-					<a href="#" class="btn" data-dismiss="modal"><?php echo JText::_('COM_TEMPLATES_TEMPLATE_CLOSE'); ?></a>
+					<button type="button" class="btn" data-dismiss="modal"><?php echo JText::_('COM_TEMPLATES_TEMPLATE_CLOSE'); ?></button>
 					<button class="btn btn-primary" type="submit"><?php echo JText::_('COM_TEMPLATES_BUTTON_RESIZE'); ?></button>
```
Comment (Quy) [inline administrator/templates/hathor/html/com_templates/template/default.php]: Move `type` before `class`

File: administrator/components/com_menus/views/menus/tmpl/default.php [hunk 9]
```diff
@@ -213,7 +213,7 @@
 								<?php endforeach; ?>
 							<?php elseif ($modMenuId) : ?>
 								<?php $link = JRoute::_('index.php?option=com_modules&task=module.add&eid=' . $modMenuId . '&params[menutype]=' . $item->menutype . '&tmpl=component&layout=modal'); ?>
-								<a class="btn btn-small btn-primary" data-toggle="modal" role="button" href="#moduleAddModal"><?php echo JText::_('COM_MENUS_ADD_MENU_MODULE'); ?></a>
+								<button class="btn btn-small btn-primary" type="button" data-toggle="modal" data-target="#moduleAddModal"><?php echo JText::_('COM_MENUS_ADD_MENU_MODULE'); ?></button>
```
Comment (Quy) [inline administrator/components/com_menus/views/menus/tmpl/default.php]: Move `type` before `class`.

File: administrator/components/com_contact/views/contacts/tmpl/default_batch_footer.php [hunk 10]
```diff
@@ -9,9 +9,9 @@
 defined('_JEXEC') or die;
 
 ?>
-<a class="btn" type="button" onclick="document.getElementById('batch-category-id').value='';document.getElementById('batch-access').value='';document.getElementById('batch-language-id').value='';document.getElementById('batch-user-id').value='';document.getElementById('batch-tag-id').value=''" data-dismiss="modal">
+<button type="button" class="btn" onclick="document.getElementById('batch-category-id').value='';document.getElementById('batch-access').value='';document.getElementById('batch-language-id').value='';document.getElementById('batch-user-id').value='';document.getElementById('batch-tag-id').value=''" data-dismiss="modal">
 	<?php echo JText::_('JCANCEL'); ?>
-</a>
+</button>
 <button class="btn btn-success" type="submit" onclick="Joomla.submitbutton('contact.batch');">
```
Comment (Quy) [inline administrator/components/com_contact/views/contacts/tmpl/default_batch_footer.php]: Move `type` before `class`

File: administrator/components/com_categories/views/categories/tmpl/default_batch_footer.php [hunk 11]
```diff
@@ -9,9 +9,9 @@
 defined('_JEXEC') or die;
 
 ?>
-<a class="btn" type="button" onclick="document.getElementById('batch-category-id').value='';document.getElementById('batch-access').value='';document.getElementById('batch-language-id').value=''" data-dismiss="modal">
+<button type="button" class="btn" onclick="document.getElementById('batch-category-id').value='';document.getElementById('batch-access').value='';document.getElementById('batch-language-id').value=''" data-dismiss="modal">
 	<?php echo JText::_('JCANCEL'); ?>
-</a>
+</button>
 <button class="btn btn-success" type="submit" onclick="Joomla.submitbutton('category.batch');">
```
Comment (Quy) [inline administrator/components/com_categories/views/categories/tmpl/default_batch_footer.php]: Move `type` before `class`

File: administrator/components/com_menus/views/items/tmpl/default_batch_footer.php [hunk 12]
```diff
@@ -11,9 +11,9 @@
 $clientId  = $this->state->get('filter.client_id');
 $menuType = JFactory::getApplication()->getUserState('com_menus.items.menutype');
 ?>
-<a class="btn" type="button" onclick="document.getElementById('batch-menu-id').value='';document.getElementById('batch-access').value='';document.getElementById('batch-language-id').value=''" data-dismiss="modal">
+<button type="button" class="btn" onclick="document.getElementById('batch-menu-id').value='';document.getElementById('batch-access').value='';document.getElementById('batch-language-id').value=''" data-dismiss="modal">
 	<?php echo JText::_('JCANCEL'); ?>
-</a>
+</button>
 <?php if ((strlen($menuType) && $menuType != '*' && $clientId == 0) || ($published > 0 && $clientId == 1)): ?>
 	<button class="btn btn-success" type="submit" onclick="Joomla.submitbutton('item.batch');">
```
Comment (Quy) [inline administrator/components/com_menus/views/items/tmpl/default_batch_footer.php]: Move `type` before `class`

File: administrator/components/com_messages/views/messages/view.html.php [hunk 13]
```diff
@@ -90,10 +90,10 @@ protected function addToolbar()
 			0,
 			'',
 			'',
-			'<button class="btn" type="button" data-dismiss="modal" aria-hidden="true">'
+			'<button class="btn" type="button" data-dismiss="modal">'
```
Comment (Quy) [inline administrator/components/com_messages/views/messages/view.html.php]: Move `type` before `class`

File: administrator/components/com_messages/views/messages/view.html.php [hunk 14]
```diff
@@ -90,10 +90,10 @@ protected function addToolbar()
...
 			. JText::_('JCANCEL')
 			. '</button>'
-			. '<button class="btn btn-success" type="button" data-dismiss="modal" aria-hidden="true"'
+			. '<button class="btn btn-success" type="button" data-dismiss="modal"'
```
Comment (Quy) [inline administrator/components/com_messages/views/messages/view.html.php]: Move `type` before `class`

File: administrator/templates/hathor/html/com_banners/tracks/default.php [hunk 15]
```diff
@@ -129,9 +129,9 @@
 				'url'         => JRoute::_('index.php?option=com_banners&amp;view=download&amp;tmpl=component'),
 				'width'       => '100%',
 				'height'      => '300px',
-				'footer'      => '<a class="btn" data-dismiss="modal" type="button"'
+				'footer'      => '<button type="button" class="btn" data-dismiss="modal"'
 						. ' onclick="jQuery(\'#downloadModal iframe\').contents().find(\'#closeBtn\').click();">'
-						. JText::_('COM_BANNERS_CANCEL') . '</a>'
+						. JText::_('COM_BANNERS_CANCEL') . '</button>'
 						. '<button class="btn btn-success" type="button"'
```
Comment (Quy) [inline administrator/templates/hathor/html/com_banners/tracks/default.php]: Move `type` before `class`

File: administrator/modules/mod_multilangstatus/tmpl/default.php [hunk 16]
```diff
@@ -23,12 +23,14 @@
 ?>
 
 <div class="btn-group multilanguage">
-	<a data-toggle="modal"
-		href="#multiLangModal"
+	<button
+		type="button"
+		data-toggle="modal"
```
Comment (Quy) [inline administrator/modules/mod_multilangstatus/tmpl/default.php]: Add `class="btn"`

Comment (SharkyKZ) [reply administrator/modules/mod_multilangstatus/tmpl/default.php]: Reverting this one because of styling issues.

**Comment by Quy** [comment]
I did a search on `<button class=` and found more to move `type` before `class`.

**Comment by SharkyKZ** [comment]
There are also more buttons with `type` missing. But keep that for a separate PR.

**Comment by Quy** [comment]
Why is `Security.md` in this PR?

**Comment by SharkyKZ** [comment]
EOL issue. Closing this now. Going to break it down into smaller pieces.

**Pull request closed without merging**